1-intro/intermediate.ipynb

```diff
@@ -0,0 +1,354 @@
+   "metadata": {},
+   "source": [
+    "### List extension\n",
+    "You are probably familiar with the `.append()` method for lists which lets you attach an element to the end of that list. If you want to attach a sequence to a list, instead of repeatedly appending you can extend instead. `.extend()` works similarly to `.append()` except it takes an entire sequence. Using it can make your code shorter and clearer."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "97be28c8-d024-4d6a-9f21-b595beb27161",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "list1 = [1, 2, 3]\n",
+    "list2 = [4, 5, 6]\n",
+    "\n",
+    "# The slow and verbose way\n",
+    "for elem in list2:\n",
+    "    list1.append(elem)\n",
+    "print(list1)\n",
+    "\n",
+    "list1 = [1, 2, 3]\n",
+    "# Instead do\n",
+    "list1.extend(list2)\n",
+    "print(list1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5da5e676-8406-4e84-92ca-a77f2b02a3fd",
+   "metadata": {},
+   "source": [
+    "### Functional programming\n",
+    "Python contains a couple of functions that allow a different style of programming. These functions are `map()` and `filter()`.\n",
+    "`map()` applies a specified function to each item in an iterable and returns an iterable of the results. It's a convenient way to transform data in an iterable without using explicit loops.\n",
+    "`filter()` filters elements from an iterable based on a specified condition (function) and returns an iterable containing only the elements that meet the condition. It's useful for selecting items that match a specific criterion."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f240a4c0-08a8-4ef0-bd3b-d7010eb416e4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def square(x):\n",
+    "    return x**2\n",
+    "    \n",
+    "numbers = [1, 2, 3, 4, 5]\n",
+    "\n",
+    "squared = list(map(square, numbers))\n",
+    "print(squared)\n",
+    "\n",
+    "def even(x):\n",
+    "    return x % 2 == 0\n",
+    "\n",
+    "evens = list(filter(even, numbers))\n",
+    "print(evens)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "874c990d-f7b3-4849-aa3a-f33696f49218",
+   "metadata": {},
+   "source": [
+    "### Opening files\n",
+    "Paths to files should be defined using `pathlib`. `pathlib` also offers methods such as `Path.open()`, which improves consistency and can improve readability over the `open` built-in."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d550c47b-f5a3-4c24-bdb4-00d9d9247b2d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Instead of\n",
+    "with open(\"my_file\", \"w\") as f:\n",
+    "    # Your code here\n",
+    "    pass\n",
+    "\n",
+    "# Use instead\n",
+    "from pathlib import Path\n",
+    "\n",
+    "with Path(\"my_file\").open(\"w\") as f:\n",
+    "    # Your code here\n",
+    "    pass"
```

Comment:
I don't think importing `pathlib` just to use `Path.open()` is more convenient than using the builtin `open()`. Much better examples would be something showcasing useful properties such as [`Path.parent`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.PurePath.parent) or [`Path.name`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.PurePath.name), or maybe something using [`Path.mkdir(parents=True, exist_ok=True)`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.Path.mkdir) or [`Path.read_text()`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.Path.read_text).

Reply:
[`Ruff rule PTH123`](https://docs.astral.sh/ruff/rules/builtin-open/) says to use `Path.open()` instead of `open()`.
Including some of those other methods here seems like a good idea to me.

Reply:
My point was that if I had code that was not using `pathlib` already, then using `Path.open()` instead of the built-in `open()` would be _less_ convenient because it would require an additional import. Preferring `Path.open()` only makes sense if the code is using `pathlib` anyways.